src/compute/src/render/reduce.rs

```diff
@@ -345,42 +352,69 @@ where
 
                 // We expect not to have any negative multiplicities, but are not 100% sure it will
                 // never happen so for now just log an error if it does.
-                for (val, cnt) in input.iter() {
-                    if *cnt < 0 {
-                        warn!("[customer-data] Negative accumulation in ReduceCollation: {val:?} with count {cnt:?} in dataflow {debug_name}");
-                        soft_assert_or_log!(false, "Negative accumulation in ReduceCollation");
-                    }
-                }
-
-                for ((reduction_type, row), _) in input.iter() {
-                    match reduction_type {
-                        ReductionType::Accumulable => {
-                            accumulable = row.iter();
-                        }
-                        ReductionType::Hierarchical => {
-                            hierarchical = row.iter();
+                if input.iter().any(|(_val, cnt)| cnt < &0) {
+                    for (val, cnt) in input.iter() {
+                        // XXX: This reports user data, which we perhaps should not do!
```

Comment:
nit: maybe ping @ggnall regarding that...

Reply:
@ggnall To clarify, the behavior to output customer data to Sentry in the case of invalid accumulations in reductions is something that was introduced long before this PR. So this PR does not change this situation. The extra logging was requested since it is pretty hard to track these errors down when they do occur. The Slack thread for reference is [this one](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1670493167314799).

Fixing the approach taken in `src/compute/src/render/reduce.rs` is the subject of epic #17178. It is not straightforward to come up with a solution, but my personal preference would be to convert these error logs into SQL-level errors (design doc forthcoming).

Reply:
Regarding the reporting of user data, see discussion at https://materializeinc.slack.com/archives/CM7ATT65S/p1668169080723029